messagequeue/messagequeue.go

```diff
@@ -0,0 +1,216 @@
+	out     bsmsg.BitSwapMessage
+	network bsnet.BitSwapNetwork
+	wl      *wantlist.ThreadSafe
+
+	sender bsnet.MessageSender
+
+	refcnt int
+
+	work chan struct{}
+	done chan struct{}
+}
+
+func New(p peer.ID, network bsnet.BitSwapNetwork) MessageQueue {
+	return &msgQueue{
+		done:    make(chan struct{}),
+		work:    make(chan struct{}, 1),
+		wl:      wantlist.NewThreadSafe(),
+		network: network,
+		p:       p,
+		refcnt:  1,
+	}
+}
+
+func (mq *msgQueue) RefIncrement() {
+	mq.refcnt++
+}
+
+func (mq *msgQueue) RefDecrement() bool {
+	mq.refcnt--
+	return mq.refcnt > 0
+}
+
+func (mq *msgQueue) AddMessage(entries []*bsmsg.Entry, ses uint64) {
+	var work bool
+	mq.outlk.Lock()
+	defer func() {
+		mq.outlk.Unlock()
+		if !work {
+			return
+		}
+		select {
+		case mq.work <- struct{}{}:
+		default:
+		}
+	}()
+
+	// if we have no message held allocate a new one
+	if mq.out == nil {
+		mq.out = bsmsg.New(false)
+	}
+
+	// TODO: add a msg.Combine(...) method
+	// otherwise, combine the one we are holding with the
+	// one passed in
+	for _, e := range entries {
+		if e.Cancel {
+			if mq.wl.Remove(e.Cid, ses) {
+				work = true
+				mq.out.Cancel(e.Cid)
+			}
+		} else {
+			if mq.wl.Add(e.Cid, e.Priority, ses) {
+				work = true
+				mq.out.AddEntry(e.Cid, e.Priority)
+			}
+		}
+	}
+}
+
+func (mq *msgQueue) Startup(ctx context.Context, initialEntries []*wantlist.Entry) {
+
+	// new peer, we will want to give them our full wantlist
+	fullwantlist := bsmsg.New(true)
+	for _, e := range initialEntries {
+		for k := range e.SesTrk {
+			mq.wl.AddEntry(e, k)
+		}
+		fullwantlist.AddEntry(e.Cid, e.Priority)
+	}
+	mq.out = fullwantlist
+	mq.work <- struct{}{}
+
+	go mq.runQueue(ctx)
+}
+
+func (mq *msgQueue) Shutdown() {
+	close(mq.done)
+}
+func (mq *msgQueue) runQueue(ctx context.Context) {
+	for {
+		select {
+		case <-mq.work: // there is work to be done
+			mq.doWork(ctx)
+		case <-mq.done:
+			if mq.sender != nil {
+				mq.sender.Close()
+			}
+			return
+		case <-ctx.Done():
+			if mq.sender != nil {
+				mq.sender.Reset()
+			}
+			return
+		}
+	}
+}
+
+func (mq *msgQueue) doWork(ctx context.Context) {
+	// grab outgoing message
+	mq.outlk.Lock()
+	wlm := mq.out
+	if wlm == nil || wlm.Empty() {
+		mq.outlk.Unlock()
+		return
+	}
+	mq.out = nil
+	mq.outlk.Unlock()
+
+	// NB: only open a stream if we actually have data to send
+	if mq.sender == nil {
+		err := mq.openSender(ctx)
+		if err != nil {
+			log.Infof("cant open message sender to peer %s: %s", mq.p, err)
+			// TODO: cant connect, what now?
+			return
+		}
+	}
+
+	// send wantlist updates
+	for { // try to send this message until we fail.
+		err := mq.sender.SendMsg(ctx, wlm)
+		if err == nil {
+			return
+		}
+
+		log.Infof("bitswap send error: %s", err)
+		mq.sender.Reset()
+		mq.sender = nil
+
+		select {
+		case <-mq.done:
+			return
+		case <-ctx.Done():
+			return
+		case <-time.After(time.Millisecond * 100):
+			// wait 100ms in case disconnect notifications are still propogating
```

Comment:
@Stebalien do you think this is still necessary with the somewhat-recent fixes to connection notifications?

Reply:
I don't think notifications will affect this. This is a race between writes failing because the connection is closed and the stream muxer externalizing the fact that the connection is closed.

Before opening a stream, the swarm will check `conn.IsClosed` and then choose another connection if it's closed. However, writes to streams may start failing before the stream muxer notices that he connection is closed (we usually "notice" because we get an error on *read*, not *write*).